### ISSUE TITLE
bug(lib): Explicitly stating wrong return type of `Promise.then` without arguments results in no error

### ISSUE THREAD

**Issue by iluha168** [microsoft/TypeScript#63244]
### 🔎 Search Terms

- silenced Promise resolution type mismatch
- Promise generic explicit no error
- no error when Promise type is specified explicitly
- generic inside optional argument

### 🕗 Version & Regression Information

TS v3.3.3 correctly detects the error, which is the earliest Playground version. All other Playground versions silence the error.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=6.0.0-dev.20260312#code/MYewdgzgLgBFCm0BcMAKAnEBbAlheAPNOjmAOYB8MAvGprvgHTqIgA2AbvABQCMAlIygALeGG78A3EA

### 💻 Code

```ts
const test: Promise<string> = Promise.resolve(1).then();
```

### 🙁 Actual behavior

No TypeScript errors.

### 🙂 Expected behavior

I expected a similar error to what TS v3.3.3 yields:
> Type `Promise<number>` is not assignable to type `Promise<string>`. Type `number` is not assignable to type `string`.

**Comment by jcalz** [comment]
So maybe `then()` should be typed more like

```ts
interface Promise<T> {
    then<T1 = T, T2 = never>(
        onfulfilled?: ((value: T) => T1 | PromiseLike<T1>) | undefined | null,
        onrejected?: ((reason: any) => T2 | PromiseLike<T2>) | undefined | null
    ): Promise<NoInfer<T1 | T2>>;
}
```

with [`NoInfer<>`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-4.html#the-noinfer-utility-type)?

**Comment by iluha168** [comment]
The return value hint usually is useful, and adding `NoInfer` is a huge breaking change:
```ts
const specificPromiseType: Promise<1> = Promise.resolve().then(() => 1)
   // ^? Promise<1>
const notSpecificPromiseType          = Promise.resolve().then(() => 1)
   // ^? Promise<number>
```
Moreover, generics can be set manually:
```ts
const test = Promise.resolve(5).then<"string">()
	// ^? Promise<NoInfer<string>>
```

Perhaps we could utilize overloads and split this definition into two - one with function `onfulfilled`, and the other - with non-function `onfulfilled`? (No need to split `onrejected`, for the same reason `Promise.catch` does not have [this](https://github.com/microsoft/TypeScript/issues/63244) issue - its default type is `never`, any other generic would be less specific.)

```ts
interface PromiseLike<T> {
    then<TResult1 = T, TResult2 = never>(
        onfulfilled: (value: T) => TResult1 | PromiseLike<TResult1>,
        onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null
    ): PromiseLike<TResult1 | TResult2>;
    then<EasierMigrationGenericIGuess = T, TResult2 = never>(
        onfulfilled?: undefined | null,
        onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null
    ): PromiseLike<T | TResult2>;
}
// And the same for Promise<T> except the return type is also a Promise
```

The sample code now results in this error [[Playground]](https://www.typescriptlang.org/play/?noLib=true&ts=5.9.3#code/JYOwLgpgTgZghgYwgAgApQPYFtgGcIAywA1hADwAqAfMgN4BQyTyYAFhCJQEoS4CuAGzABGZAF5kFADSSe-IQCZxyEBABu0KgApGzPRhAxBMYAIEQAJgC5kWtXAF8INigEpxNCnMEjkAHzRMHHwiUm5eH2EqKV09JgMoCAArCARICwB+Gy0tRLhcAxs4EABPdzFPb0V-QOw8QhJyLwjFKncAvhALCBNVCxqQQQFY5FcbdDqQxvD5XwDm2YUqAG4Rtg4yAFF84GgAWWAAcyg4MGADAHEOaGAEAEkLp1xcZWlZFrAlCVUNKG0R-SGYymcyZGydbq9SwDIYxOLMBLJVLpLK2XIQfKFZDFMoed6LGoTYINMILHxLdrICE9UDQgKDMwjMa1YmhJo1MmtVYAX3o9G6CAEcESyAQBlwYGQWBKRPq4yC9TZZEGWAARpp6GKQBKWLwwPLJiTyBKoKBDjQJNLZfgAHTrEBaVzLIA):
```ts
const test: Promise<string> = Promise.resolve(1).then();
/*
Type 'Promise<string | number>' is not assignable to type 'Promise<string>'.
  Type 'string | number' is not assignable to type 'string'.
    Type 'number' is not assignable to type 'string'
*/
```
The error description is not perfect, but it does result in expected behavior if `onrejected` is split too. [[Playground]](https://www.typescriptlang.org/play/?noLib=true&ts=5.9.3#code/JYOwLgpgTgZghgYwgAgApQPYFtgGcIAywA1hADwAqAfMgN4BQyTyYAFhCJQEoS4CuAGzABGZAF5kFADSSe-IQCZxyEBABu0KgApGzPRhAxBMYAIEQAJgC5kWtXAF8INigEpxNCnMEjkAHzRMHHwiUm5eH2EqKV09JgMoCAArCARIa1tEuFwDGzgQAE93MU9vRX9A7DxCEnIvCMUqWORXG3QqkNrw+V8A+p6FKgBuZrYObsjlaWQAUWzgaABZYABzKDgwYAMAcQ5oYAQASW2nXFxlVQ0obWb9Q2NTcwy7BycXYtKG3srgmrD+yLRW7MBLJVLpAD8Nj4IAsEBMqgsFRAggEzVaP2qoTqZREw1G7E4c1wCygyzWGy2IF2qigB2Op3OEmmAPKEkumh0cRB9wEJjMlmhsPhoEsyNRMW58RAiRSaUFmQg2VyyHyRQ8si+SgC7V+2ImjXRbSCWK6FAqrLAgxGejGRPmS1W602Oz2dKOJ14TMkMmJpPJzqpNP2HsZSnZ6k5wOlRj5j0sUOQMLhCLFARRZklUtBcshQpToqR6dRRsxnX++IAvvR6HCEAI4IlkAgDLgwMgsAVddVjR0-uQUVgAEaaegtkBtli8MC9vVdNt0kArGgSTvd-AAOjtWlcQyAA)

**Comment by Andarist** [comment]
Another instance of when `.then`'s typing affects some scenario negatively can be seen here: https://github.com/microsoft/TypeScript/issues/62071

**Comment by paigefletcher1282-code** [comment]
Yes https://github.com/microsoft/TypeScript/issues/62071  this hasn't yet
to be reproduced

**Comment by iluha168** [comment]
@Andarist @paigefletcher1282-code Interestingly enough, my overload idea does indeed solve the specific case from [#62071](https://github.com/microsoft/TypeScript/issues/62071)'s sample code. [(Playground)](https://www.typescriptlang.org/play/?noLib=true&ts=5.9.3#code/C4TwDgpgBAgg7gQwJbAgEwDwBUB8UC8UWUEAHqgHZoDOUACgE4D2AtktRBkhQGYQNQASngD8QqAC4oFCADd+UAPSKiAeQAiqgFBbuqBjwQBjaI1bsIAYSYVqwBgFcjwJgIDeWqF6gMICNDYANiBQYMwuoJBSZmwcGAgUIDgA3DreUAiBgdgk5BBUtL7+QSEOFADWFExwFADaALpQAD5QDTgAFLKZDhDUUlgAlNHMsZxuUAC0RQEUwa10UNxQ5RAgTDxE9VLwyKiYWLV09SlQAL4pad6+1EyB8u1D9CMWGLJMSGgX6de38tgdXUCPX6jxiLx2KHQ-y+V16v04uE63Qg-WaT3McVwoOecQhe2hqVOOjQECMgQQvigXQEYI4wwxVhsdkczlcqV0FH0hhM6NG-ygHnSwAAFvlsIJeg5AsAAIwEIgAGiIEuoUuAACZ5TJ5AwOp50t4bDwpTwkFl0FIkUCUUQBgQ8FgVWq5S1aQindKZTgFfqDVAbL4AFakvYiS3tIo3ChSBIgO34B0ejVot3iyXS9U4O0tMok00yNBoihSwK+qDYhlp1WetGO9MamFeEVigCiCGoSH4AFkkABzBgIYBIGwAcXy-CQRgAkiOetRaIQsEq69Xk4Rtfw9X6vEaTWbAugw1BcxB8+giyWfdv-RQgyHD+HIzYY4l44n65rXTj3R+s2iT2ehYtMWWRlhWfLEC0K5qpmhLEqS5KUsaFDOMOFBQL2EDAAAcg4LAPPSfLFiwABGm6pCSZIUtAyGoTYGFYQAyvYBG8i8zLcL2FyUYhNFlHR6EOBwWAOGAB7tMAokHlItTEWRDBKhxFC9vUjxvB8qRGEywChPKboAHQ-HcEDtA0Az6c2FDtA89psRw+mZIEpmYTheEPEqLnMQwDyqeZlntEJEAiWJEADKkQA)
```ts
declare function getNum(): Promise<number>;
declare function getStr(): Promise<string>;
declare function useTuple(tuple: [number, string]): void;
const p = Promise.resolve([]).then(() => Promise.all([getNum(), getStr()])).then(useTuple);
/*                                                                            ^
                                 (method) Promise<[number, string]>.then<void, never>(
                                   onfulfilled: (value: [number, string]) => void | Promise<void>,
                                   onrejected?: ((reason: any) => Promise<never>) | null | undefined
                                 ): Promise<void> (+1 overload)
*/
```
I have, however, encountered a bug with `Awaited` - in its lib implementation, `T extends object & { then(onfulfilled: infer F, ...args: infer _): any; }` infers `F` to `null | undefined`. That should be addressed in a proposed fix... _Edit: including current `then` type definition as a third overload fixes `infer`, and therefore, `Awaited` as well._
